README.md
[![Version](https://vsmarketplacebadge.apphb.com/version/bitlang.cobol.svg)](https://marketplace.visualstudio.com/items?itemName=bitlang.cobol) [![Installs](https://vsmarketplacebadge.apphb.com/installs-short/bitlang.cobol.svg)](https://marketplace.visualstudio.com/items?itemName=bitlang.cobol) [![Downloads](https://vsmarketplacebadge.apphb.com/downloads-short/bitlang.cobol.svg)](https://marketplace.visualstudio.com/items?itemName=bitlang.cobol) [![Rating](https://vsmarketplacebadge.apphb.com/rating-star/bitlang.cobol.svg)](https://marketplace.visualstudio.com/items?itemName=bitlang.cobol)


# COBOL Source editing for Visual Studio Code

This extension was originally aimed at providing syntax highlighting for COBOL but overtime is has been extended to provide syntax highlighting for other related languages/file formats such JCL, PL/I and Micro Focus directive files and Micro Focus Unit Test Reports.

## What can I expect from this extension

This extension is certainly not an Integrated Development Environment for COBOL.   It does provide enough functionality for the casual developer to view, edit and compile source code (via tasks & problem matchers).

## Does this extension include a language server protocol for COBOL

Visual Studio Code has two ways of providing language extensions, the first is provided by the use of a [language server protocol](https://langserver.org/) and the second is via a set of Visual Studio Code APIs.

This extension uses the Visual Studio Code APIs because the Language Server Protocol did not exist when this extension was created.

## What platform can it be used on?

Everywhere Visual Studio Code works.

## Code colorization for COBOL, PL/I and JCL:

 ![sieve_jcl](https://raw.githubusercontent.com/spgennard/vscode_cobol/master/images/screenshot_three.png)

## IntelliSense example:

![perform_add](https://raw.githubusercontent.com/spgennard/vscode_cobol/master/images/perform_add.gif)

## Breadcrumb support:

![breadcrumbs](https://raw.githubusercontent.com/spgennard/vscode_cobol/master/images/breadcrumb.png)

## Outline support:

![outline](https://raw.githubusercontent.com/spgennard/vscode_cobol/master/images/outline.png)

## Go to definition:

![gotodef](https://raw.githubusercontent.com/spgennard/vscode_cobol/master/images/gotodef.gif)

## Peek definition:

![peekdef](https://raw.githubusercontent.com/spgennard/vscode_cobol/master/images/peekdef.gif)

## Keybinds

| Keys                     |                          Description                            |
|--------------------------|:---------------------------------------------------------------:|
| ctrl+alt+p               |                    Go to procedure division                     |
| ctrl+alt+w               |                  Go to working-storage section                  |
| ctrl+alt+d               | Go to data division (or working-storage section if not present) |
| ctrl+alt+,               |             Go backwards to next section/division               |
| ctrl+alt+.               |            Go forward to next next section/division             |
| f12 or ctrl+click        |                     Go to copybook/file                         |
| ctrl+hover               |                     Peek head of copybook or symbol/field       |
| right mouse/peek         |            Peek copybook without opening the file)              |

## Settings

- COBOL tab stops can be changed by editing the ```coboleditor.tabstops``` setting.
- Extensions used for *Go to copybook*, can be changed by editing the ```coboleditor.copybookexts``` settings.
- Directories used for *Go to copybook*, can be changed by editing the ```coboleditor.copybookdirs``` settings.

## Tasks

Visual Studio code can be setup to build your COBOL source code.

### Task: Using MsBuild

MsBuild based projects can be consumed as build task, allowing navigation to error/warnings when they occur.

Below is an example of *build* task that uses *mycobolproject.sln*.

```json
{
    "version": "2.0.0",
    "tasks": [ {
            "label": "build",
            "type": "shell",
            "command": "msbuild",
            "args": [
                "/property:GenerateFullPaths=true",
                "/t:build",
                "mycobolproject.sln"
            ],
            "group": {
                "kind": "build",
                "isDefault": true
            },
            "presentation": {
                "reveal": "always"
            },
            "problemMatcher": "$mfcobol-msbuild"
        }
    ]
}
```

### Task: Single file compile using Micro Focus COBOL - ERRFORMAT(3)

The example below shows you how you can create a single task to compile one program using the `cobol` command.

For Net Express/Server Express compilers use the "$mfcobol-errformat3-netx-sx" problem matcher as although the directive ERRFORMAT"3" is used, the compiler output error format is slightly different.

```json
{
    "label": "mf cobol (single file)",
    "command": "cobol",
    "args": [
        "${file}",
        "noint",
        "nognt",
        "noobj",
        "noquery",
        "errformat(3)",
        "COPYPATH($COBCPY;${workspaceFolder}\\CopyBooks;${workspaceFolder}\\CopyBooks\\Public)",
        ";"
    ],
    "group": {
        "kind": "build",
        "isDefault": true
    },
    "options": {
        "cwd": "${workspaceRoot}"
    },
    "presentation": {
        "echo": true,
        "reveal": "never",
        "focus": true,
        "panel": "dedicated"
    },
    "problemMatcher": "$mfcobol-errformat3"
}
```

### Task: Single file compile using Micro Focus COBOL - ERRFORMAT(2)

The example below shows you how you can create a single task to compile one program using the `cobol` command.

For Net Express/Server Express compilers use the "$mfcobol-errformat2-netx-sx" problem matcher as although the directive ERRFORMAT"2" is used, the compiler output error format is slightly different.

```json
{
    "label": "mf cobol (single file)",
    "command": "cobol",
    "args": [
        "${file}",
        "noint",
        "nognt",
        "noobj",
        "noquery",
        "errformat(2)",
        "COPYPATH($COBCPY;${workspaceFolder}\\CopyBooks;${workspaceFolder}\\CopyBooks\\Public)",
        ";"
    ],
    "group": {
        "kind": "build",
        "isDefault": true
    },
    "options": {
        "cwd": "${workspaceRoot}"
    },
    "presentation": {
        "echo": true,
        "reveal": "never",
        "focus": true,
        "panel": "dedicated"
    },
    "problemMatcher": [ "$mfcobol-errformat2", "$mfcobol-errformat2-copybook" ]
}
```

### Task: Single file compile using GnuCOBOL/OpenCOBOL/COBOL-IT

The example below shows you how you can create a single task to compile one program using the `cobc` command.

This example is for GnuCOBOL 1-2.x, for GnuCOBOL use $gnucobol3-cob

```json
{
    "version": "2.0.0",
    "tasks": [
        {
            "label": "gnucobol - cobc (single file)",
            "type": "shell",
            "command": "cobc",
            "args": [
                "-fsyntax-only",
                "-I${workspaceFolder}\\CopyBooks",
                "-I${workspaceFolder}\\CopyBooks\\Public",
                "${file}"
            ],
            "problemMatcher" : "$gnucobol-cobc"
        }
    ]
}
```

### Task: Single file compile using ACUCOBOL-GT

The example below shows you how you can create a single task to compile one program using the `ccbl32` command.

```json
{
    "version": "2.0.0",
    "tasks": [
        {
            "label": "acu cobol - ccbl32 (single file)",
            "type": "shell",
            "command": "%ACUCOBOL%\\bin\\ccbl32",
            "args": [
                "-Sp", "${workspaceFolder}\\CopyBooks",
                "-Sp", "${workspaceFolder}\\CopyBooks\\Public",
                "${file}"
            ],
            "windows": {
                "options": {
                    "env": {
                        "ACUCOBOL" : "C:\\extend10.1.1\\AcuGT"
                    }
                }
            },
            "problemMatcher" : [ "$acucobol-warning-ccbl", "$acucobol-ccbl" ]
        }
    ]
}
```

### Task: Breakdown of problem matchers

| Product and Version                           | Tools                                                            | Problem matcher(s)                                                     |
|-----------------------------------------------|------------------------------------------------------------------|------------------------------------------------------------------------|
| COBOL-IT                                      | *cobc*                                                           | $cobolit-cobc                                                          |
| COBOL-IT                                      | *cobc* for errors/notes                                          | $cobolit-error-cobc + $cobolit-note-cobc                               |
| open-cobol 1-1.5                              | *cobc*                                                           | $opencobol-cobc                                                        |
| open-cobol 1-1.5                              | *cobc* for warnings/errors/notes                                 | $opencobol-warning-cobc + $opencobol-error-cobc + $opencobol-note-cobc |
| ACU-COBOLGT                                   | *ccbl* for errors/warnings                                       | $acucobol-ccbl + $acucobol-warning-ccbl                                |
| GnuCOBOL 1-2                                  | *cobc*                                                           | $gnucobol-cobc                                                         |
| GnuCOBOL 3                                    | *cobc*                                                           | $gnucobol3-cobc                                                        |
| GnuCOBOL 3                                    | *cobc* for warnings/errors/notes                                 | $gnucobol3-warning-cobc + $gnucobol3-error-cobc + $gnucobol3-note-cobc |
| Micro Focus COBOL Net Express/Server Express  | *cob* or *cobol.exe* + ERRFORMAT"2"                              | $mfcobol-errformat2-netx-sx                                            |
|                                               | *cob* or *cobol.exe* + ERRFORMAT"2" for errors in copybooks      | +mfcobol-errformat2-copybook-netx-sx                                   |
|                                               | *cob* or *cobol.exe* + ERRFORMAT"3"                              | $mfcobol-errformat3-netx-sx                                            |
|                                               | *cob* or *cobol.exe* + ERRFORMAT"3" for information              | +mfcobol-errformat3-info                                               |
| Micro Focus Visual COBOL/Enterprise Developer | *msbuild*                                                        | $mfcobol-msbuild                                                       |
|                                               | *cob* or *cobol.exe* + ERRFORMAT"3"                              | $mfcobol-errformat3                                                    |
|                                               | *cob* or *cobol.exe* + ERRFORMAT"3" / filename extract with PATH | $mfcobol-errformat3-basefn                                             |
|                                               | *cob* or *cobol.exe* + ERRFORMAT"2"                              | $mfcobol-errformat2                                                    |
|                                               | *cob* or *cobol.exe* + ERRFORMAT"2" for errors in copybooks      | $mfcobol-errformat2-copybook                                           |

NOTE: Problem matchers can be stacked in your task definition.   It is recommended that any "-copybook", "-info", "-note" and similar problem matcher are included before problem matchers without this suffix.

## Remote development using containers

If your main development is Micro Focus Visual COBOL/Enterprise Developer you may have access to base images that provide the compiler and its tools.

If you do, all that is required is another image that contains extra tools and a devcontainer.json to configure its use.

The following ```Dockerfile``` is an example on how you can extend your existing base image with java configured, ant, git and lsb tools.

This example uses the SLES 15.1 base images using Visual COBOL 5.0.

You may need to tweak the ```FROM``` clause in the Dockerfile and if you use a different platform or product version, the ```zypper``` will also require a change too if a different platform is used (different commands eg: yum, microdnf etc..).

Dockerfile:

```Dockerfile
FROM microfocus/vcdevhub:sles15_5.1_x64_login

USER root

ENV JAVA_HOME=/usr/java/default
ENV PATH=${JAVA_HOME}/bin:${PATH}

ENV COBDIR=${MFPRODBASE}

ENV PATH=${COBDIR}/bin:${PATH}
ENV LD_LIBRARY_PATH=${COBDIR}/lib:${LD_LIBRARY_PATH}

ENV ANT_HOME=/opt/microfocus/VisualCOBOL/remotedev/ant/apache-ant-1.9.9
ENV PATH=${ANT_HOME}/bin:${PATH}

RUN zypper --non-interactive install  --no-recommends lsb-release git
```

devcontainer.json:

```json
{
	// See https://aka.ms/vscode-remote/devcontainer.json for format details.
	"name": "Visual COBOL",

	// Update the 'dockerFile' property if you aren't using the standard 'Dockerfile' filename.
	"dockerFile": "Dockerfile",

	// The optional 'runArgs' property can be used to specify additional runtime arguments.
	"runArgs": [
		//  Uncomment the next line if you want to use Docker from the container. See the docker-in-docker definition for details.
		// "-v","/var/run/docker.sock:/var/run/docker.sock",

		// Uncomment the next two lines if you will use a ptrace-based debugger like C++, Go, and Rust
		"--cap-add=SYS_PTRACE",
		"--security-opt", "seccomp=unconfined"
	],

	// Uncomment the next line if you want to publish any ports.
	// "appPort": [],

	// Uncomment the next line if you want to add in default container specific settings.json values
	// "settings":  { "workbench.colorTheme": "Quiet Light" },

	// Uncomment the next line to run commands after the container is created.
	// "postCreateCommand": "uname -a",

	// Add the IDs of any extensions you want installed in the array below.
	"extensions": [
		"bitlang.cobol"
	]
}
```

### Workspace Setup

Visual Studio Code workspaces are not "projects" but do allow you keep your source in one place.

Things to consider:

 - If you have copybooks, you should change the ```coboleditor.copybookdirs``` settings to setup where the extension can find your copybooks
  - If use the *COPY* verb with this extension, you may also need to adjust the ```coboleditor.copybookexts``` settings (json array).
  - If you want the extension to understand the contents of your copybook before you access it, then turn on the ```coboleditor.parse_copybooks_for_references``` settings (json array) to allow the extension to look inside the copybook references


### COBOL Linter

The COBOL linter included with the extension performs two functions, the first function is to identify sections/paragraphs that are not used and the second is to apply any "house" standards.

The section/parapgraph linter by default generates warning message but if you prefer the messages to be marked as information, you change the ```coboleditor.linter_mark_as_information``` boolean setting, for example:

 ```json
 "coboleditor.linter_mark_as_information": true
 ```

The house standards are applied to fields in various sections, each section is named and a rule defined as a regular expression to enforce.

For example to enforce all working-storage items must start with ws and local-storage ls, you can use:

```json
"coboleditor.linter_house_standards_rules": [
        "file=.*",
        "thread-local=ls.*",
        "working-storage=ws.*",
        "object-storage=.*",
        "local-storage=.*",
        "linkage=.*",
        "communication=.*",
        "report=.*",
        "screen=.*",
    ]
```

### Metadata caching location

#### Recommend use

By default the metadata caching is turned off but it can be turned on via the ```coboleditor.cache_metadata``` setting.

This setting, allows you to specific where the metdata caching files are stored.

The recommendation setting for the metadata caching is the ```workspace``` option.   This setting creates a .vscode_cobol directory that contains all the metadata files within the workspace folder.

This directory should be excluded from source code control system.

Example configuration:

```json
    "coboleditor.cache_metadata": "workspace"
```

#### Advanced use

If the user wants to specific where the metdata caching directory is to be located, the ```user_defined_directory``` setting can be used.


CHANGELOG.md

# Changelog

Changelog for spgennard vscode_cobol.

## Unreleased
### GitHub [#216](https://github.com/spgennard/vscode_cobol/issues/216) How to disable a single entry (ideally only for a specific source) from being shown in the problem pane?  

**Start to include docs for linter - Fix #216**


[0179c56c6b5001e](https://github.com/spgennard/vscode_cobol/commit/0179c56c6b5001e) *2020-11-10 23:55:03*


### GitHub [#243](https://github.com/spgennard/vscode_cobol/issues/243) Issue with metadata in storagepath  

**Issue with metadata in storagepath Fix #243**

 * In order resolve the storage creep and permissions issues with the 
 * use of storagepath, the hard decision to remove it and been done. 
 * - remove storagepath and repace it with user_defined_directory 
 * more work/testing to be done before it is released 

[688204568aa904d](https://github.com/spgennard/vscode_cobol/commit/688204568aa904d) *2020-11-11 18:48:11*


## 6.9.20
### GitHub [#229](https://github.com/spgennard/vscode_cobol/issues/229) add acucobol-listing  

**- add acucobol-listing Fix #229**

 * - remove &quot;COBOL:&quot; and add cat field 

[8a901e9d62d6e0d](https://github.com/spgennard/vscode_cobol/commit/8a901e9d62d6e0d) *2020-11-06 22:31:48*


## 6.9.19
### GitHub [#229](https://github.com/spgennard/vscode_cobol/issues/229) add acucobol-listing  

**add acucobol-listing Fix #229**

 * - fix observed issues with various tmLanguage files (copyright entity) 

[ede84ab0cbea7de](https://github.com/spgennard/vscode_cobol/commit/ede84ab0cbea7de) *2020-11-05 22:11:23*


## 6.9.18
### GitHub [#189](https://github.com/spgennard/vscode_cobol/issues/189) Add support of coboldoc  

**Add support of coboldoc - Fix #189**

 * - Second part, comment out the support 
 * Schedule cleanup next month just incase. 

[e6ae98bb8470944](https://github.com/spgennard/vscode_cobol/commit/e6ae98bb8470944) *2020-11-02 20:50:15*

**Add support of coboldoc - Fix #189**

 * - Support for COBOLDoc has been abandoned until it is more mature 

[c137fa606879251](https://github.com/spgennard/vscode_cobol/commit/c137fa606879251) *2020-11-02 20:43:47*


### GitHub [#235](https://github.com/spgennard/vscode_cobol/issues/235) MF matcher problem with COBCH038 name XYZ undeclared, line 878 (first usage)  

**MF matcher problem with COBCH038 name XYZ undeclared, line 878 (first usage) #235**


[ec7459528b6dd5c](https://github.com/spgennard/vscode_cobol/commit/ec7459528b6dd5c) *2020-11-04 21:06:10*


## 6.9.0
### GitHub [#187](https://github.com/spgennard/vscode_cobol/issues/187) Metadata entries (like AUTHOR and SOURCE-COMPUTER) are not terminated by a newline  

**Metadata entries (like AUTHOR and SOURCE-COMPUTER) are not terminated by a newline Fix #187**

 * - tweak language to handle line numbers &amp; move to using an 
 * embedded approach to the listing syntax 

[de366f628a0e993](https://github.com/spgennard/vscode_cobol/commit/de366f628a0e993) *2020-10-15 22:16:24*


### GitHub [#222](https://github.com/spgennard/vscode_cobol/issues/222) partial patch - Problem matcher for Pro*COBOL  

**partial patch - Problem matcher for Pro*COBOL Fix #222**

 * - test included &amp; changed 
 * - split problem matches out (think I will rename this) 

[f9476ee008895a0](https://github.com/spgennard/vscode_cobol/commit/f9476ee008895a0) *2020-10-16 23:21:37*


### GitHub [#224](https://github.com/spgennard/vscode_cobol/issues/224) FR: provide outline for &quot;COBOL_LISTFILE&quot;  

**FR: provide outline for "COBOL_LISTFILE" Fix #224**


[cea61e25a322862](https://github.com/spgennard/vscode_cobol/commit/cea61e25a322862) *2020-10-15 22:21:58*


### GitHub [#225](https://github.com/spgennard/vscode_cobol/issues/225) Support Pro*COBOL listing  

**Support Pro*COBOL listing Fix #225**


[4cfc966e36b5418](https://github.com/spgennard/vscode_cobol/commit/4cfc966e36b5418) *2020-10-15 22:51:54*


## 6.8.2
### GitHub [#218](https://github.com/spgennard/vscode_cobol/issues/218) How to debug &quot;no definition found&quot; for copybooks?  

**How to debug "no definition found" for copybooks? Fix #218**

 * - the letters &#39;IN&#39; in a copybook caused the def provider to miss the 
 * copybook 

[3624946165a232c](https://github.com/spgennard/vscode_cobol/commit/3624946165a232c) *2020-10-02 17:53:38*


## 6.8.1
### GitHub [#215](https://github.com/spgennard/vscode_cobol/issues/215) Extension causes high cpu load  

**Extension causes high cpu load #215**

 * - drop back to exist sync, as looking at &#39;access&#39;, it does the 
 * same but via an exception for the handling, so exist looks 
 * to be better 

[6495184fb6d98e5](https://github.com/spgennard/vscode_cobol/commit/6495184fb6d98e5) *2020-10-01 22:19:08*


## 6.8.0
### GitHub [#215](https://github.com/spgennard/vscode_cobol/issues/215) Extension causes high cpu load  

**Extension causes high cpu load #215**

 * - flip over to using fs.access to see if we get a better performance profile 

[f3df393ea728707](https://github.com/spgennard/vscode_cobol/commit/f3df393ea728707) *2020-10-01 22:06:40*


## 6.7.44
### GitHub [#215](https://github.com/spgennard/vscode_cobol/issues/215) Extension causes high cpu load  

**Extension causes high cpu load #215**

 * - reduce the times used fsStat is used by using a tuple 

[38e5a3b98bbd378](https://github.com/spgennard/vscode_cobol/commit/38e5a3b98bbd378) *2020-09-25 07:33:04*


## 6.7.43
### GitHub [#208](https://github.com/spgennard/vscode_cobol/issues/208) Feature Request: Add config setting to allow finding definitions in _any_ workspace file  

**Add config setting to allow finding definitions in _any_ workspace file #208**

 * - wire warning messages into diagnostic engine 
 * (unable to find copybook defined by hint) 

[a704f90e458561c](https://github.com/spgennard/vscode_cobol/commit/a704f90e458561c) *2020-09-24 23:14:05*


## 6.7.41
### GitHub [#199](https://github.com/spgennard/vscode_cobol/issues/199) Do we need to tweak the pre_parse_line_limit?  


src/configuration.ts
'use strict';

import { workspace } from 'vscode';
import { ICOBOLSettings, COBOLSettings, outlineFlag } from './iconfiguration';
import * as path from 'path';
import { isDirectory } from './extension';
import { CacheDirectoryStrategy } from './externalfeatures';


export class VSCOBOLConfiguration {
    private static config: ICOBOLSettings = new COBOLSettings();

    public static init(): ICOBOLSettings {
        const vsconfig = VSCOBOLConfiguration.config;

        vsconfig.experimental_features = getBoolean('experimental_features', false);
        vsconfig.enable_tabstop = getBoolean('enable_tabstop', false);
        vsconfig.ignorecolumn_b_onwards = getBoolean('ignorecolumn_b_onwards', false);
        vsconfig.copybooks_nested = getBoolean('copybooks_nested', false);
        vsconfig.fuzzy_variable_search = getBoolean('fuzzy_variable_search', false);
        vsconfig.outline = isOutlineEnabled();
        vsconfig.copybookdirs = getCopybookdirs_defaults(vsconfig.invalid_copybookdirs);
        vsconfig.pre_parse_line_limit = getPreParseLineLimit();
        vsconfig.copybookexts = getCopybookExts();
        vsconfig.program_extensions = getProgram_extensions();
        vsconfig.tabstops = getTabStops();
        vsconfig.linter = getBoolean('linter', false);
        vsconfig.line_comment = getBoolean("line_comment", false);
        vsconfig.fileformat_strategy = getFileformatStrategy();
        vsconfig.enable_data_provider = getBoolean('enable_data_provider', true);
        vsconfig.disable_unc_copybooks_directories = getBoolean('disable_unc_copybooks_directories', false);
        vsconfig.intellisense_include_unchanged = getBoolean("intellisense_include_unchanged", true);
        vsconfig.intellisense_include_camelcase = getBoolean("intellisense_include_camelcase", false);
        vsconfig.intellisense_include_uppercase = getBoolean("intellisense_include_uppercase", false);
        vsconfig.intellisense_include_lowercase = getBoolean("intellisense_include_lowercase", false);
        vsconfig.intellisense_item_limit = getIntellisense_item_limit();
        vsconfig.process_metadata_cache_on_start = getBoolean("process_metadata_cache_on_start", false);
        vsconfig.cache_metadata = getcache_metadata();
        vsconfig.cache_metadata_time_limit = getNumber("cache_metadata_time_limit", 30000);
        vsconfig.cache_metadata_max_directory_scan_depth = getNumber("cache_metadata_max_directory_scan_depth", 32);
        vsconfig.cache_metadata_show_progress_messages = getBoolean("cache_metadata_show_progress_messages", false);
        vsconfig.parse_copybooks_for_references = getBoolean("parse_copybooks_for_references", false);
        vsconfig.workspacefolders_order = getWorkspacefolders_order();
        vsconfig.linter_unused_paragraphs_or_sections = getBoolean("linter_unused_paragraphs_or_sections", true);
        vsconfig.linter_house_standards = getBoolean("linter_house_standards", true);
        vsconfig.linter_house_standards_rules = getlinter_house_standards_rules();
        vsconfig.linter_mark_as_information = getBoolean("linter_mark_as_information", true);
        vsconfig.ignore_unsafe_extensions = getBoolean("ignore_unsafe_extensions", false);
        vsconfig.coboldoc_workspace_folder = getCoboldoc_workspace_folder();
        vsconfig.process_scanner_hints_embedded_in_comments = getBoolean("process_scanner_hints_embedded_in_comments", false);
        vsconfig.process_scanner_hint_token_for_source_dependencies = getprocess_scanner_hint_token_for_source_dependencies();
        vsconfig.process_metadata_cache_on_file_save = getBoolean("process_metadata_cache_on_file_save", false);
        vsconfig.cache_metadata_user_directory = getString("cache_metadata_user_directory", "");
        vsconfig.editor_maxTokenizationLineLength = workspace.getConfiguration('editor').get<number>("maxTokenizationLineLength",20000);
        vsconfig.sourceview = getBoolean("sourceview", false);
        vsconfig.sourceview_include_jcl_files = getBoolean("sourceview_include_jcl_files", true);
        vsconfig.sourceview_include_hlasm_files = getBoolean("sourceview_include_hlasm_files", true);
        vsconfig.sourceview_include_pli_files = getBoolean("sourceview_include_pli_files", true);
        vsconfig.sourceview_include_doc_files = getBoolean("sourceview_include_doc_files", true);
        vsconfig.sourceview_include_script_files = getBoolean("sourceview_include_script_files", true);
        return vsconfig;
    }

    public static get(): ICOBOLSettings {
        if (VSCOBOLConfiguration.config.init_required) {
            VSCOBOLConfiguration.init();
            VSCOBOLConfiguration.config.init_required = false;
        }
        return VSCOBOLConfiguration.config;
    }


    public static isOnDiskCachingEnabled(): boolean {
        const config = VSCOBOLConfiguration.get();
        const cacheStrat = config.cache_metadata;
        if (cacheStrat === CacheDirectoryStrategy.Off) {
            return false;
        }

        return true;
    }
}

function getBoolean(configSection: string, defaultValue: boolean): boolean {
    const editorConfig = workspace.getConfiguration('coboleditor');
    let expEnabled = editorConfig.get<boolean>(configSection);
    if (expEnabled === undefined || expEnabled === null) {
        expEnabled = defaultValue;
    }
    return expEnabled;
}

function getString(configSection: string, defaultValue: string): string {
    const editorConfig = workspace.getConfiguration('coboleditor');
    let strValue = editorConfig.get<string>(configSection);
    if (strValue === undefined || strValue === null) {
        strValue = defaultValue;
    }
    return strValue;
}

function getNumber(configSection: string, defaultValue: number): number {
    const editorConfig = workspace.getConfiguration('coboleditor');
    let lineLimit = editorConfig.get<number>(configSection);
    if (lineLimit === undefined || lineLimit === null) {
        lineLimit = defaultValue;
    }
    return lineLimit;
}

function getPreParseLineLimit(): number {
    const editorConfig = workspace.getConfiguration('coboleditor');
    let lineLimit = editorConfig.get<number>('pre_parse_line_limit');
    if (lineLimit === undefined || lineLimit === null) {
        lineLimit = 25;
    }
    return lineLimit;
}

function getCoboldoc_workspace_folder(): string {
    const editorConfig = workspace.getConfiguration('coboleditor');
    const coboldoc_folder = editorConfig.get<string>('coboldoc_workspace_folder');
    if (coboldoc_folder === undefined || coboldoc_folder === null) {
        return "coboldoc";
    }
    return coboldoc_folder;
}

function getcache_metadata(): CacheDirectoryStrategy {
    const editorConfig = workspace.getConfiguration('coboleditor');
    let cacheDirStrategy = editorConfig.get<string>('cache_metadata');

    if (cacheDirStrategy === undefined || cacheDirStrategy === null) {
        return CacheDirectoryStrategy.Off;
    }

    cacheDirStrategy = cacheDirStrategy.toLowerCase();
    if (cacheDirStrategy === "off") {
        return CacheDirectoryStrategy.Off;
    }

    switch (cacheDirStrategy) {
        case 'workspace': return CacheDirectoryStrategy.Workspace;
        case 'user_defined_directory': return CacheDirectoryStrategy.UserDefinedDirectory;
        case 'off': return CacheDirectoryStrategy.Off;
    }
    return CacheDirectoryStrategy.Off;
}


function getprocess_scanner_hint_token_for_source_dependencies(): string {
    const editorConfig = workspace.getConfiguration('coboleditor');
    let hintToken = editorConfig.get<string>('process_scanner_hint_token_for_source_dependencies');
    if (hintToken === undefined || hintToken === null) {
        hintToken = "source-dependency";
    }
    return hintToken;
}

function getIntellisense_item_limit(): number {
    const editorConfig = workspace.getConfiguration('coboleditor');
    let itemLimit = editorConfig.get<number>('intellisense_item_limit');
    if (itemLimit === undefined || itemLimit === null) {
        itemLimit = 0;
    }
    return itemLimit;
}

function getFileformatStrategy(): string {
    const editorConfig = workspace.getConfiguration('coboleditor');
    const fileStrat = editorConfig.get<string>('fileformat_strategy');

    if (fileStrat === undefined || fileStrat === null) {
        return "normal";
    }

    return fileStrat;
}


function isOutlineEnabled(): outlineFlag {
    const editorConfig = workspace.getConfiguration('coboleditor');
    const outlineEnabled = editorConfig.get('outline');
    if (outlineEnabled === undefined || outlineEnabled === null) {
        return outlineFlag.On;
    }

    switch (outlineEnabled) {
        case "on": return outlineFlag.On;
        case "off": return outlineFlag.Off;
        case "partial": return outlineFlag.Partial;
        case "skeleton": return outlineFlag.Skeleton;
    }
    return outlineFlag.On;
}

const DEFAULT_COPYBOOK_DIR: string[] = [];

function expandEnvVars(startEnv: string): string {
    let complete = false;
    let env: string = startEnv;

    while (complete === false) {
        const indexOfEnv = env.indexOf("${env:");
        if (indexOfEnv === -1) {
            complete = true;
        } else {
            const lenOfValue = env.indexOf("}") - (indexOfEnv + 6);
            const envValue = env.substr(6 + indexOfEnv, lenOfValue);
            const left = env.substr(0, indexOfEnv);
            const right = env.substr(1 + env.indexOf("}"));
            env = left + process.env[envValue] + right;
        }
    }

    return env;
}

function getCopybookdirs_defaults(invalidSearchDirectory: string[]): string[] {
    const editorConfig = workspace.getConfiguration('coboleditor');
    let dirs = editorConfig.get<string[]>('copybookdirs');
    if (!dirs || (dirs !== null && dirs.length === 0)) {
        dirs = DEFAULT_COPYBOOK_DIR;
    }

    const extraDirs: string[] = [];

    for (let dirpos = 0; dirpos < dirs.length; dirpos++) {
        let dir = dirs[dirpos];

        /* remove ${workspaceFolder} */
        dir = expandEnvVars(dir);

        if (dir.startsWith("${workspaceFolder}")) {
            dir = dir.replace("${workspaceFolder}", "").trim();

            // remove / or \ forward
            if (dir.startsWith('/') || dir.startsWith('\\')) {
                dir = dir.substr(1).trim();
            }
        }

        // ignore empty elements
        if (dir.length !== 0) {
            if (dir.startsWith("$")) {
                const e = process.env[dir.substr(1)];
                if (e !== undefined && e !== null) {
                    e.split(path.delimiter).forEach(function (item) {
                        if (item !== undefined && item !== null && item.length > 0) {
                            if (isDirectory(item)) {
                                extraDirs.push(item);
                            } else {
                                invalidSearchDirectory.push(item);
                            }
                        }
                    });
                } else {
                    invalidSearchDirectory.push(dir);
                }
            } else {
                if (dir !== ".") {
                    extraDirs.push(dir);
                }
            }
        }
    }

    return extraDirs;
}

const DEFAULT_COPYBOOK_EXTS = ["cpy", "CPY"];
const DEFAULT_PROGRAM_EXTS = ["cbl", "CBL"];

function getCopybookExts(): string[] {
    const editorConfig = workspace.getConfiguration('coboleditor');
    let extensions = editorConfig.get<string[]>('copybookexts');
    if (!extensions || (extensions !== null && extensions.length === 0)) {
        extensions = DEFAULT_COPYBOOK_EXTS;
    }
    extensions.push("");
    return extensions;
}

function getProgram_extensions(): string[] {
    const editorConfig = workspace.getConfiguration('coboleditor');
    let extensions = editorConfig.get<string[]>('program_extensions');
    if (!extensions || (extensions !== null && extensions.length === 0)) {
        extensions = DEFAULT_PROGRAM_EXTS;
    }
    return extensions;
}


const DEFAULT_RULER = [0, 7, 11, 15, 19, 23, 27, 31, 35, 39, 43, 47, 51, 55, 59, 63, 67, 71, 75, 79];

function getTabStops(): number[] {
    const editorConfig = workspace.getConfiguration('coboleditor');
    let tabStops = editorConfig.get<number[]>('tabstops');
    if (!tabStops || (tabStops !== null && tabStops.length === 0)) {
        tabStops = DEFAULT_RULER;
    }
    return tabStops;
}

function getWorkspacefolders_order(): string[] {
    const editorConfig = workspace.getConfiguration('coboleditor');
    let dirs = editorConfig.get<string[]>('workspacefolders_order');
    if (!dirs || (dirs !== null && dirs.length === 0)) {
        dirs = [];
    }
    return dirs;
}


function getlinter_house_standards_rules(): string[] {
    const editorConfig = workspace.getConfiguration('coboleditor');
    let standards = editorConfig.get<string[]>('linter_house_standards_rules');
    if (!standards || (standards !== null && standards.length === 0)) {
        standards = [];
    }
    return standards;
}